plugins/parsers/csv/parser.go

```diff
@@ -0,0 +1,156 @@
+package csv
+
+import (
+	"bytes"
+	"encoding/csv"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+)
+
+type CSVParser struct {
+	MetricName      string
+	Header          bool
+	Delimiter       string
+	DataColumns     []string
+	TagColumns      []string
+	FieldColumns    []string
+	NameColumn      string
+	TimestampColumn string
+	TimestampFormat string
+	DefaultTags     map[string]string
+}
+
+func (p *CSVParser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	r := bytes.NewReader(buf)
+	csvReader := csv.NewReader(r)
+	csvReader.FieldsPerRecord = len(p.DataColumns)
+	if p.Delimiter != "" {
+		runeStr := []rune(p.Delimiter)
+		if len(runeStr) > 1 {
+			return nil, fmt.Errorf("delimiter must be a single character, got: %v", p.Delimiter)
+		}
+		csvReader.Comma = runeStr[0]
+	}
+
+	//if there is a header and nothing in DataColumns
+	//set DataColumns to names extracted from the header
+	if p.Header && len(p.DataColumns) == 0 {
+		header, err := csvReader.Read()
+		if err != nil {
+			return nil, err
+		}
+		p.DataColumns = header
+
+	} else if p.Header {
+		//if there is a header and DataColumns is specified, just skip header
+		csvReader.Read()
+
+	} else if !p.Header && len(p.DataColumns) == 0 {
+		//if there is no header and no DataColumns, that's an error
+		return nil, fmt.Errorf("there must be a header if `csv_data_columns` is not specified")
+	}
+
+	table, err := csvReader.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, record := range table {
+		m, err := p.parseRecord(record)
+		if err != nil {
+			return metrics, err
+		}
+		metrics = append(metrics, m)
+	}
+	return metrics, nil
+}
+
+//does not use any information in header and assumes DataColumns is set
```

Comment:
If a comment is directly above an exported function, start it with `// FunctionName ...`